File: moxy/src/main/java/com/omegar/mvp/MvpProcessor.java
package com.omegar.mvp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.omegar.mvp.presenter.PresenterField;
import com.omegar.mvp.presenter.PresenterType;
import com.omegar.mvp.presenter.InjectPresenter;

/**
 * Date: 18-Dec-15
 * Time: 13:16
 *
 * @author Yuri Shmakov
 * @author Alexander Blinov
 */
public class MvpProcessor {
	public static final String PRESENTER_BINDER_SUFFIX = "$$PresentersBinder";
	public static final String PRESENTER_BINDER_INNER_SUFFIX = "Binder";
	public static final String VIEW_STATE_SUFFIX = "$$State";
	public static final String VIEW_STATE_PROVIDER_SUFFIX = "$$ViewStateProvider";

	/**
	 * <p>1) Generates tag for identification MvpPresenter</p>
	 * <p>2) Checks if presenter with tag is already exist in {@link PresenterStore}, and returns it</p>
	 * <p>3) If {@link PresenterStore} doesn't contain MvpPresenter with current tag, {@link PresenterField} will create it</p>
	 *
	 * @param <Delegated>    type of delegated
	 * @param target         object that want injection
	 * @param presenterField info about presenter from {@link InjectPresenter}
	 * @param delegateTag    unique tag @return MvpPresenter instance
	 */
	private <Delegated> MvpPresenter<? super Delegated> getMvpPresenter(Delegated target,
																		PresenterField<Delegated> presenterField,
																		String delegateTag) {
		PresenterStore presenterStore = MvpFacade.getInstance().getPresenterStore();

		PresenterType type = presenterField.getPresenterType();
		String tag;
		if (type == PresenterType.LOCAL) {
			tag = delegateTag + "$" + presenterField.getTag(target);
		} else {
			tag = presenterField.getTag(target);
		}

		//noinspection unchecked
		MvpPresenter<? super Delegated> presenter = presenterStore.get(tag);
		if (presenter != null) {
			return presenter;
		}

		//noinspection unchecked
		presenter = (MvpPresenter<? super Delegated>) presenterField.providePresenter(target);

		if (presenter == null) {
			return null;
		}

		presenter.setPresenterType(type);
		presenter.setTag(tag);
		presenterStore.add(tag, presenter);

		return presenter;
	}

	/**
	 * <p>Gets presenters {@link List} annotated with {@link InjectPresenter} for view.</p>
	 * <p>See full info about getting presenter instance in {@link #getMvpPresenter}</p>
	 *
	 * @param delegated   class contains presenter
	 * @param delegateTag unique tag
	 * @param <Delegated> type of delegated
	 * @return presenters list for specifies presenters container
	 */
	<Delegated> List<MvpPresenter<? super Delegated>> getMvpPresenters(Delegated delegated,
																	   String delegateTag,
																	   List<PresenterField<Delegated>> customPresenterFields) {
		List<MvpPresenter<? super Delegated>> presenters = new ArrayList<>();

		List<Object> presenterBinders = getPresenterBinders(delegated);

		for (Object presenterBinderObject : presenterBinders) {
			//noinspection unchecked
			PresenterBinder<Delegated> presenterBinder = (PresenterBinder<Delegated>) presenterBinderObject;
			List<PresenterField<Delegated>> presenterFields = presenterBinder.getPresenterFields();

			for (PresenterField<Delegated> presenterField : presenterFields) {
				handlePresenterField(delegated, delegateTag, presenters, presenterField);
			}
		}

		// handle custom presenter fields
		if (customPresenterFields != null) {
			for (PresenterField<Delegated> presenterField : customPresenterFields) {
				handlePresenterField(delegated, delegateTag, presenters, presenterField);
			}
		}

		return presenters;
	}

	private <Delegated> List<Object> getPresenterBinders(Delegated delegated) {
		if (!hasMoxyReflector()) return Collections.emptyList();

		@SuppressWarnings("unchecked")
		Class<? super Delegated> aClass = (Class<Delegated>) delegated.getClass();
		List<Object> presenterBinders = null;

		while (aClass != Object.class && presenterBinders == null) {
			presenterBinders = MoxyReflector.getPresenterBinders(aClass);

			aClass = aClass.getSuperclass();
		}

		if (presenterBinders == null || presenterBinders.isEmpty()) {
			return Collections.emptyList();
		}
		return presenterBinders;
	}

	private <Delegated> void handlePresenterField(Delegated delegated, String delegateTag, List<MvpPresenter<? super Delegated>> presenters, PresenterField<Delegated> presenterField) {
		PresentersCounter presentersCounter = MvpFacade.getInstance().getPresentersCounter();

		MvpPresenter<? super Delegated> presenter = getMvpPresenter(delegated, presenterField, delegateTag);

		if (presenter != null) {
			presentersCounter.injectPresenter(presenter, delegateTag);
			presenters.add(presenter);
			presenterField.bind(delegated, presenter);
		}
	}

	private static Boolean hasMoxyReflector = null;

	// Check is it have generated MoxyReflector without usage of reflection API
	private static boolean hasMoxyReflector() {
		if (hasMoxyReflector != null) {
			return hasMoxyReflector;
		}

		try {
			new MoxyReflector();

			hasMoxyReflector = true;
		} catch (NoClassDefFoundError error) {
			hasMoxyReflector = false;
		}

		return hasMoxyReflector;
	}
}

File: moxy/src/main/java/com/omegar/mvp/MvpProcessor.kt
package com.omegar.mvp

import com.omegar.mvp.MvpFacade.presenterStore
import com.omegar.mvp.MvpFacade.presentersCounter
import com.omegar.mvp.presenter.PresenterField
import com.omegar.mvp.presenter.PresenterType
import kotlin.reflect.KClass

/**
 * Date: 18-Dec-15
 * Time: 13:16
 *
 * @author Yuri Shmakov
 * @author Alexander Blinov
 */
object MvpProcessor {

    private var hasMoxyReflector: Boolean = try {
        MoxyReflector
        true
    } catch (error: NoClassDefFoundError) {
        false
    }

    /**
     *
     * 1) Generates tag for identification MvpPresenter
     *
     * 2) Checks if presenter with tag is already exist in [PresenterStore], and returns it
     *
     * 3) If [PresenterStore] doesn't contain MvpPresenter with current tag, [PresenterField] will create it
     *
     * @param <Delegated>    type of delegated
     * @param target         object that want injection
     * @param presenterField info about presenter from [InjectPresenter]
     * @param delegateTag    unique tag @return MvpPresenter instance
    </Delegated> */
    private fun <Delegated : Any, Presenter : MvpPresenter<*>> getMvpPresenter(
        target: Delegated,
        presenterField: PresenterField<Delegated, Presenter>,
        delegateTag: String
    ): Presenter {
        return getOrCreateMvpPresenter(
            delegateTag = delegateTag,
            presenterType = presenterField.presenterType,
            presenterClass = presenterField.presenterClass
        ) {
            presenterField.providePresenter(target)
        }
    }

    @Suppress("UNCHECKED_CAST")
    fun <Presenter : MvpPresenter<*>> getOrCreateMvpPresenter(
        delegateTag: String,
        presenterType: PresenterType,
        presenterClass: KClass<*>,
        presenterFactory: () -> Presenter
    ): Presenter {
        val presenterTag = generatePresenterTag(presenterType, delegateTag, presenterClass)
        return presenterStore[presenterTag] as Presenter? ?: presenterFactory().also { presenter ->
            presenter.presenterType = presenterType
            presenter.presenterTag = presenterTag
        }
    }

    /**
     * Gets presenters [List] annotated with [InjectPresenter] for view.
     *
     * See full info about getting presenter instance in [.getMvpPresenter]
     *
     * @param delegated   class contains presenter
     * @param delegateTag unique tag
     * @param <Delegated> type of delegated
     * @return presenters list for specifies presenters container
    </Delegated> */
    fun <Delegated : Any, P : MvpPresenter<*>> getMvpPresenters(
        delegated: Delegated,
        delegateTag: String,
        customPresenterFields: List<PresenterField<Delegated, P>>?
    ): List<P> {
        return getPresenterBinders<Delegated, P>(delegated::class)
            .flatMap { it.presenterFields }
            .plus(customPresenterFields.orEmpty())
            .map { presenterField ->
                getMvpPresenter(delegated, presenterField, delegateTag).also { presenter ->
                    presentersCounter.injectPresenter(presenter, delegateTag)
                    presenterField.bind(delegated, presenter)
                }
            }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <Delegated : Any, P : MvpPresenter<*>> getPresenterBinders(delegatedClass: KClass<out Delegated>):
            List<PresenterBinder<Delegated, P>> {
        if (!hasMoxyReflector) return emptyList()

        val presenterBinders = MoxyReflector.getPresenterBinders(delegatedClass) as List<PresenterBinder<Delegated, P>>?

        return presenterBinders.orEmpty()
    }

    /**
     * @return generated tag in format: &lt;parent_delegate_tag&gt; &lt;delegated_class_full_name&gt;$MvpDelegate@&lt;hashCode&gt;
     *
     * example: SampleFragment$MvpDelegate@32649b0
     */
    fun generateDelegateTag(delegatedClass: KClass<*>, delegateClass: KClass<*>, uniqueKey: Int): String {
        return delegatedClass.simpleName + "$" + delegateClass.simpleName + "@" + Integer.toHexString(uniqueKey)
    }

    fun generatePresenterTag(type: PresenterType, delegateTag: String, presenterClass: KClass<*>): String {
        return when (type) {
            PresenterType.LOCAL -> delegateTag + presenterClass.qualifiedName
            else -> presenterClass.qualifiedName.orEmpty()
        }
    }

}